app/gui2/src/util/ast/opr.ts

```diff
@@ -1,73 +1,104 @@
+import { mapIterator } from 'lib0/iterator'
 
 /** An operand of one of the applications inside `GeneralOprApp` */
-export type GeneralOperand =
-  | Operand
+export type GeneralOperand<HasIdMap extends boolean = true> =
+  | Operand<HasIdMap>
   // A part of `GeneralOprApp`, consisting of lhs and first `statements` of applications.
-  | { type: 'partOfGeneralOprApp'; oprApp: GeneralOprApp; statements: number }
+  | { type: 'partOfGeneralOprApp'; oprApp: GeneralOprApp<HasIdMap>; statements: number }
+
+export type OperatorChain<HasIdMap extends boolean = true> = AstExtended<
+  Tree.OprApp | Tree.OperatorBlockApplication,
+  HasIdMap
+>
 
 /** A structure unifying API of OprApp and OperatorBlockApplication */
-export class GeneralOprApp {
-  lhs: Tree | null
+export class GeneralOprApp<HasIdMap extends boolean = true> {
+  lhs: AstExtended<Tree, HasIdMap> | null
   apps: {
-    opr: Result<Token.Operator, MultipleOperatorError>
-    expr: Tree | null
+    opr: AstExtended<Token.Operator, HasIdMap> | null
+    expr: AstExtended<Tree, HasIdMap> | null
   }[]
 
-  constructor(ast: Tree.OprApp | Tree.OperatorBlockApplication) {
-    this.lhs = ast.lhs ?? null
-    if (ast.type === Tree.Type.OprApp) {
-      const rhs = ast.rhs ?? null
-      this.apps = [{ opr: ast.opr, expr: rhs }]
+  constructor(ast: OperatorChain<HasIdMap>) {
+    this.lhs = ast.tryMap((t) => t.lhs) ?? null
+    if (ast.isTree(Tree.Type.OprApp)) {
+      const rhs = ast.tryMap((t) => t.rhs) ?? null
+      const opr = ast.tryMap((t) => (t.opr.ok ? t.opr.value : undefined)) ?? null
+      this.apps = [{ opr, expr: rhs }]
     } else {
-      function* nonEmptyLines(ast: Tree.OperatorBlockApplication) {
-        for (const line of ast.expressions) {
-          if (line.expression == null) continue
-          yield { opr: line.expression.operator, expr: line.expression.expression }
+      const blockApplication = ast as AstExtended<Tree.OperatorBlockApplication, HasIdMap>
+      const operators = blockApplication.tryMapIter((ast) =>
+        [...ast.expressions]
+          .flatMap((line) => (line.expression ? [line.expression] : []))
+          .map((expr) => (expr.operator.ok ? expr.operator.value : null))
+          .values(),
+      )
+      const exprs = blockApplication.mapIter((ast) =>
+        [...ast.expressions]
+          .flatMap((line) => (line.expression ? [line.expression] : []))
+          .map((expr) => expr.expression)
+          .values(),
+      )
+      function* zip<T, U>(left: Iterable<T>, right: Iterable<U>): Generator<[T, U]> {
+        const leftIterator = left[Symbol.iterator]()
+        const rightIterator = right[Symbol.iterator]()
+
+        while (true) {
+          const leftResult = leftIterator.next()
+          const rightResult = rightIterator.next()
+
+          if (leftResult.done || rightResult.done) {
+            break
+          }
+
+          yield [leftResult.value, rightResult.value]
         }
       }
```

Comment:
I guess it could be in some more accessible place.

Reply:
perhaps `util/iterable.ts`

Reply:
Done